sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2151,7 +2151,7 @@ class Analyzer(
             // TODO: skip null handling for not-nullable primitive inputs after we can completely
             // trust the `nullable` information.
             val inputsNullCheck = nullableTypes.zip(inputs)
-              .filter { case (nullable, _) => !nullable }
+              .filter { case (nullable, expr) => !nullable && !expr.isInstanceOf[KnownNotNull] }
```

Comment:
let us use the original way? in the PR https://github.com/apache/spark/pull/21851
```Scala
val needsNullCheck = ... 
```